hasktorch/src/Torch/Typed/Functional.hs

```diff
@@ -6063,20 +6173,60 @@ upsample_bicubic2d _align_corners _input = unsafePerformIO $ (ATen.cast3 ATen.Ma
 -- upsample_nearest1d :: Tensor device dtype shape -> Int -> Tensor device dtype shape
 -- upsample_nearest1d _input _output_size = unsafePerformIO $ (ATen.cast2 ATen.Managed.upsample_nearest1d_tl) _input _output_size
 
--- | Applies a 2D bicubic upsampling to an input signal composed of several input channels.
+-- | Applies a 2D nearest neighbor upsampling to an input signal composed of several input channels.
 --
 -- >>> (dtype &&& shape) $ upsample_nearest2d @3 @5 (ones :: CPUTensor 'D.Float '[2,3,2,2])
 -- (Float,[2,3,3,5])
+
 upsample_nearest2d ::
-  forall w h shape dtype device.
+  forall h w shape dtype device.
   (KnownNat h, KnownNat w, All KnownNat shape) =>
   Tensor device dtype shape ->
   Tensor device dtype (Upsample2d shape h w)
-upsample_nearest2d _input = unsafePerformIO $ (ATen.cast2 ATen.Managed.upsample_nearest2d_tl) _input ([w, h] :: [Int])
+upsample_nearest2d _input = unsafePerformIO $ (ATen.cast2 ATen.Managed.upsample_nearest2d_tl) _input ([h, w] :: [Int])
   where
     w = natValI @w :: Int
     h = natValI @h :: Int
 
+
+-- Freeform Resizing
+interpolate2d :: 
+  forall newShape mode w h shape dtype device. 
+  (KnownShape shape, KnownNat w, KnownNat h, KnownSymbol mode) => 
+  Bool -> 
+  Tensor device 'D.Float shape -> 
+  Tensor device 'D.Float newShape
+interpolate2d alignCorners tensor    
+  | symbolVal (Proxy @mode) == "nearest" = unsafePerformIO $ (ATen.cast2 ATen.Managed.upsample_nearest2d_tl) tensor ([h, w] :: [Int])
+  | symbolVal (Proxy @mode) == "bilinear" = unsafePerformIO $ (ATen.cast3 ATen.Managed.upsample_bilinear2d_tlb) tensor ([h, w] :: [Int]) alignCorners
+  | otherwise = error "Invalid mode for interpolation"
+    where
+      w = natValI @w 
+      h = natValI @h
+ 
+type Resize2D (shape :: [Nat]) (newShape :: [Nat]) (n :: Nat) (c :: Nat) (h :: Nat) (w :: Nat) (n0 :: Nat) (c0 :: Nat) (h0 :: Nat) (w0 :: Nat) = 
+  ( 
+    KnownShape shape,
+    KnownShape newShape,
+    All KnownNat newShape,
+    4 ~ ListLength shape,
+    4 ~ ListLength newShape,
+    (n ': c ': h ': w : '[]) ~ newShape,
+    (n0 ': c0 ': h0 ': w0 ': '[]) ~ shape,
+    n ~ n0,
+    c ~ c0
+  )
+
+
+resize2d :: 
+  forall newShape mode shape dtype device n c h w n0 c0 h0 w0. 
+  (Resize2D shape newShape n c h w n0 c0 h0 w0, KnownSymbol mode, KnownDType dtype) => 
+  Bool -> 
+  Tensor device dtype shape -> 
+  Tensor device dtype newShape
+resize2d alignCorners = toDType @dtype @'D.Float . interpolate2d @newShape @mode @w @h alignCorners . toDType @'D.Float @dtype
```

Comment:
similar comments as for `interpolate2d`. I think that you may have intended `interpolate2d` as an internal method? If so, please add it in a where clause within the scope of `resize2d`.